core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/PartialHealthCheckStrategy.java

```diff
@@ -0,0 +1,202 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint.healthcheck;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.jctools.maps.NonBlockingHashSet;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * A strategy to check part of candidates by health.
+ */
+final class PartialHealthCheckStrategy implements HealthCheckStrategy {
+
+    /**
+     * Creates a new builder.
+     */
+    public static PartialHealthCheckStrategyBuilder builder() {
+        return new PartialHealthCheckStrategyBuilder();
+    }
+
+    private final Set<Endpoint> selectedCandidates;
+    private final Set<Endpoint> unhealthyCandidates;
+    private final TargetCount max;
+    private Set<Endpoint> candidates;
+
+    /**
+     * Creates a new instance.
+     */
+    PartialHealthCheckStrategy(TargetCount max) {
+        this.max = requireNonNull(max, "max");
+        selectedCandidates = new HashSet<>();
+        unhealthyCandidates = new NonBlockingHashSet<>();
+        candidates = ImmutableSet.of();
+    }
+
+    @Override
+    public void updateCandidates(List<Endpoint> candidates) {
+        requireNonNull(candidates, "candidates");
+
+        synchronized (selectedCandidates) {
+            this.candidates = ImmutableSet.copyOf(candidates);
+
+            final Set<Endpoint> removedCandidates = new HashSet<>();
+            for (Endpoint selectedCandidate : selectedCandidates) {
+                if (this.candidates.contains(selectedCandidate)) {
+                    continue;
+                }
+
+                removedCandidates.add(selectedCandidate);
+            }
+
+            removeAndSelectNewCandidates(removedCandidates);
+        }
+    }
+
+    @Override
+    public List<Endpoint> getCandidates() {
+        synchronized (selectedCandidates) {
+            return ImmutableList.copyOf(selectedCandidates);
+        }
+    }
+
+    @Override
+    public boolean updateHealth(Endpoint endpoint, double health) {
+        final double unhealthyScore = 0;
+
+        requireNonNull(endpoint, "endpoint");
+
+        if (!candidates.contains(endpoint)) {
+            unhealthyCandidates.remove(endpoint);
+            return true;
+        }
+
+        if (health > unhealthyScore) {
+            unhealthyCandidates.remove(endpoint);
+            return false;
+        }
+
+        unhealthyCandidates.add(endpoint);
+        synchronized (selectedCandidates) {
+            removeAndSelectNewCandidates(ImmutableSet.of(endpoint));
+            return true;
```

Comment:
I wonder if there's a case where the newly updated `selectedCandidates` actually contains the same candidates with the old `selectedCandidates`. Is there an efficient way to check such a case and return `false`?

Reply:
It's better! 🤗
I will make an `oldSelectedCandidates` when the `removeAndSelectNewCandidates()` is beginning, and check a change and return it by boolean type.